### PR TITLE
Fairadapt inclusion in AIF360

### PR DESCRIPTION
**Pull request by dplecko**
Hi guys,

Here is a pull request adding three files we mentioned (`fairadapt.py`, `test_fairadapt.py` and `demo_fairadapt.ipynb`).

Most importantly: the installation of the necessary R-packages using `rpy2` is currently done from the `fairadapt` class (check `fairadapt.py` line 54). I am guessing you would be looking for a better solution, i.e., that R-packages are installed together with AIF. Please just make sure, if this option is used, that we have installed:
- `ranger` version `0.13.1`
- `fairadapt` version `0.2.0`

Both of the above are now available via CRAN, which should make installing them using conda much easier!

Thank you in advance for attending to this!

Best,
Drago

### CODE REVIEW

File: aif360/algorithms/preprocessing/fairadapt.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+		self.prot_attr = prot_attr
+		self.adj_mat = adj_mat
+		self.outcome = outcome
+
+		# R packages need to run fairadapt
+		packnames = ('ranger', 'fairadapt')
+		# selectively install the missing packages
+		names_to_install = [x for x in packnames if not robjects.packages.isinstalled(x)]
+		if len(names_to_install) > 0:
+			utils.install_packages(StrVector(names_to_install))
```
[the remark this answers is not in the document]
Comment (dplecko) [reply aif360/algorithms/preprocessing/fairadapt.py:58]: this the `utils` package from base R - I have renamed this now to avoid name clashing.

File: aif360/algorithms/preprocessing/fairadapt.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+class fairadapt(BaseEstimator):
+	"""
+	Fairadapt is a pre-processing technique that can be used for fair classification
+	or fair regression (Plecko & Meinshausen, JMLR 2020).
+	References:
+		JMLR paper: https://www.jmlr.org/papers/volume21/19-966/19-966.pdf [1]
+		Github page: https://github.com/dplecko/fairadapt [2]
+	"""
+	def __init__(self,
```
[the remark this answers is not in the document]
Comment (dplecko) [reply aif360/algorithms/preprocessing/fairadapt.py:37]: I added `prot_attr_` and `groups_` - those two seemed sensible.

File: aif360/algorithms/preprocessing/fairadapt.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+import numpy as np
+from sklearn.base import BaseEstimator, MetaEstimatorMixin, clone
+from sklearn.utils.metaestimators import if_delegate_has_method
+from sklearn.utils.validation import has_fit_parameter
+
+from aif360.sklearn.utils import check_inputs, check_groups
+
+
+class fairadapt(BaseEstimator):
```
[the remark this answers is not in the document]
Comment (dplecko) [reply aif360/algorithms/preprocessing/fairadapt.py:29]: The documentation should now be fixed.

File: aif360/algorithms/preprocessing/fairadapt.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			X_fair_train (pandas.DataFrame): Transformed training data.
+			y_fair_train (array-like): Transformed labels of the training samples.
+			X_fair_test (pandas.DataFrame): Transformed test data.
+
+			, ,
+		"""
+		# merge X_train and y_train
+		df_train = pd.concat([X_train, y_train], axis = 1)
+
+		Fairadapt_R = robjects.r(
```
[the remark this answers is not in the document]
Comment (dplecko) [reply aif360/algorithms/preprocessing/fairadapt.py:78]: I think this is the easiest solution... it could done slightly differently, but would not look much better I believe.

File: aif360/algorithms/preprocessing/fairadapt.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+		# R packages need to run fairadapt
+		packnames = ('ranger', 'fairadapt')
+		# selectively install the missing packages
+		names_to_install = [x for x in packnames if not robjects.packages.isinstalled(x)]
+		if len(names_to_install) > 0:
+			utils.install_packages(StrVector(names_to_install))
+
+
+	def fit_transform(self, X_train, y_train, X_test):
```
[the remark this answers is not in the document]
Comment (dplecko) [reply aif360/algorithms/preprocessing/fairadapt.py:61]: so the `fairadapt` workflow does support separating the training and testing; however, when doing so, I would have to be attaching some S3 classes from R to a python object... and would make everything look more wild. I believe this is a good intermediate solution.

File: examples/sklearn/demo_fairadapt_sklearn.ipynb [hunk 6]
```diff
@@ -0,0 +1,424 @@
+{
```
[the remark this answers is not in the document]
Comment (dplecko) [reply examples/sklearn/demo_fairadapt_sklearn.ipynb:1]: Ok, this is now expanded with much more detail.

File: tests/sklearn/test_fairadapt.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    adj_mat.at["marital-status", ["education-num","workclass", "hours-per-week", "occupation","annual-income"]] = 1
+    adj_mat.at["education-num", ["workclass", "hours-per-week","occupation", "annual-income"]] = 1
+    adj_mat.at[["workclass", "hours-per-week", "occupation"], "annual-income"] = 1
+
+    pandas2ri.activate()
+
+    FA = fairadapt.fairadapt(prot_attr = "sex", adj_mat = adj_mat, outcome = "annual-income")
+    Xf_train, yf_train, Xf_test = FA.fit_transform(X_train, y_train, X_test)
+
+    assert isinstance(Xf_train, pd.DataFrame)
```
[the remark this answers is not in the document]
Comment (dplecko) [reply tests/sklearn/test_fairadapt.py:52]: Yep, no problem, I added a check to see that the metric improved.

**Review by nrkarthikeyan — APPROVED** [review]

**Comment by nrkarthikeyan** [comment]
**CI workflows failing at the R interface**

_@dplecko's analysis:_

Fixed several warnings, but the major thing to be solved is the following:
when trying to install fairadapt R-package using rpy2, I come across a problem because the default paths on the machine where the continuous integration is running are not writable
`WARNING  rpy2.rinterface_lib.callbacks:callbacks.py:123 R[write to console]:  'lib = "/usr/local/lib/R/site-library"' is not writable`
for this reason, the installation tries to run interactively and ultimately fails. When running locally, I doubt that this would be a problem.
I was wondering if adding fairadapt to ci.yml would solve the problem? In particular, adding it here:
`L81:        run: install.packages(c("reticulate", "rstudioapi", "testthat"))`
I am wondering if that would help? Not sure if the .libPaths() are shared between rpy2 and default R distribution on the machine.
Alternatively, the package installation could be moved somewhere else. Do you perhaps know how to specify a dependency to an rpy2-R-package directly in setup.py ?

@SSaishruthi @hoffmansc - Any thoughts?

**Comment by hoffmansc** [comment]
Maybe I'm missing something but do we not need to install R first in the `build-py` section of ci.yml? i.e.,
```
- name: Set up R
  uses: r-lib/actions/setup-r@v1
```
